provider/overrides/runner.rb

```diff
@@ -0,0 +1,179 @@
+# Copyright 2018 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'api/product'
+require 'provider/overrides/resources'
+require 'provider/overrides/validator'
+
+module Provider
+  module Overrides
+    # This runner takes an Api::Product and applies a set of Provider::Overrides::ResourceOverrides
+    # It does this by building a brand new Api::Product object from scratch, using
+    # the values from either the original Api::Product or the override values.
+    # Example usage in a provider.yaml file where you want to extend a resource
+    # description:
+    #
+    # overrides: !ruby/object:Provider::Overrides::ResourceOverrides
+    #   SomeResource: !ruby/object:Provider::MyProvider::ResourceOverride
+    #     description: '{{description}} A tool-specific description complement'
+    #     parameters:
+    #       someParameter: !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'foobar' # replaces description
+    #     properties:
+    #       someProperty: !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'foobar' # replaces description
+    #       anotherProperty.someNestedProperty:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #       anotherProperty[].someNestedPropertyInAnArray:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #   ...
+    class Runner
```

Comment:
Classes called -er are often a code smell and not really necessary. We never use Runner without immediately calling `.build()` on it precisely once, so I'm not sure why we need a class here? This could probably just be a function.

Reply:
That's a valid point. You've seen me do this design pattern before.

Here's my rationale and I'm curious what others think:

I have this large chunk of code that's split up into many functions. I wanted to split this code off into some unit that's responsible for the overall action this code achieves. In addition, I want to be able to mark most of this code as not being necessary outside of some entry-level function.

In my mind, I've mostly described a class just with no instance variables.

Reply:
Static method it is.